pandas/core/base.py

```diff
@@ -1402,7 +1402,17 @@ def memory_usage(self, deep=False):
         ),
     )
     def factorize(self, sort=False, na_sentinel=-1):
-        return algorithms.factorize(self, sort=sort, na_sentinel=na_sentinel)
+        codes, uniques = algorithms.factorize(
+            self._values, sort=sort, na_sentinel=na_sentinel
+        )
+        if isinstance(self, ABCIndexClass):
+            # use constructor instead of Index to get MultiIndex right
+            uniques = self._constructor(uniques)
+        else:
+            from pandas import Index
```

Comment:
what changed that this is needed now? Or is this fixing a bug?

Reply:
This was needed before the implementation of your previous suggestion, should be revertable, will check.